reco_utils/dataset/pandas_df_utils.py

```diff
@@ -316,53 +316,38 @@ def negative_feedback_sampler(
         3   1   0
     """
     # Get all of the users and items.
-    users = df[col_user].unique()
     items = df[col_item].unique()
-
-    # Create a dataframe for all user-item pairs
-    df_neg = user_item_pairs(
-        pd.DataFrame(users, columns=[col_user]),
-        pd.DataFrame(items, columns=[col_item]),
-        col_user,
-        col_item,
-        user_item_filter_df=df,
-    )
-    df_neg[col_label] = 0
-
-    df_pos = df.copy()
-    df_pos[col_label] = 1
-
-    df_all = pd.concat([df_pos, df_neg], ignore_index=True, sort=True)
-    df_all = df_all[[col_user, col_item, col_label]]
-
-    # Sample negative feedback from the combined dataframe.
-    df_sample = (
-        df_all.groupby(col_user)
-        .apply(
-            lambda x: pd.concat(
-                [
-                    x[x[col_label] == 1],
-                    x[x[col_label] == 0].sample(
-                        min(
-                            max(
-                                round(len(x[x[col_label] == 1]) * ratio_neg_per_user), 1
-                            ),
-                            len(x[x[col_label] == 0]),
-                        ),
-                        random_state=seed,
-                        replace=False,
-                    )
-                ],
-                ignore_index=True,
-                sort=True,
-            )
+    rng = np.random.default_rng(seed=seed)
+
+    def sample_items(user_df):
+        # Sample negative items for the data frame restricted to a specific user
+        n_u = len(user_df)
+        neg_sample_size = max(round(n_u * ratio_neg_per_user), 1)
+        # Draw (n_u + neg_sample_size) items and keep neg_sample_size of these
+        # that are not already in user_df. This requires a set difference from items_sample
+        # instead of items, which is more efficient when len(items) is large.
+        sample_size = min(n_u + neg_sample_size, len(items))
+        items_sample = rng.choice(items, sample_size, replace=False)
+        new_items = np.setdiff1d(items_sample, user_df[col_item])[:neg_sample_size]
+        new_df = pd.DataFrame(
+            data={
+                col_user: user_df[col_user][0],
+                col_item: new_items,
+                col_label: 0,
+            }
         )
+        return pd.concat([user_df, new_df], ignore_index=True)
+
+    res_df = df.copy()
```

Comment:
I think we should consider to add an `inplace` parameter, if it is True we don't copy the full df, otherwise, we copy

Reply:
I think this cannot be really an inplace function because it adds new rows to the data frame. So it involves concatenations of data frames, which inevitably create new copies.

Reply:
makes sense